crates/astria-sequencer/src/mempool.rs

```diff
@@ -117,6 +133,81 @@ impl std::hash::Hash for EnqueuedTransaction {
     }
 }
 
+#[derive(Debug)]
+pub(crate) struct TransactionExpired;
+
+impl std::fmt::Display for TransactionExpired {
+    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
+        write!(f, "Transaction expired")
+    }
+}
+
+const TX_TTL: Duration = Duration::from_secs(600); // 10 minutes 
+const REMOVAL_CACHE_SIZE: NonZeroUsize = unsafe { NonZeroUsize::new_unchecked(4096) }; // TODO make configuration variable 
+const TX_REMOVAL_TTL: Duration = Duration::from_secs(30); // 30 seconds
+
+/// `RemovalCache` is a cache used for signaling to `CometBFT` when a
+/// transaction can be removed from the `CometBFT` mempool outside of the
+/// `CheckTx` checks.
+///
+/// This is useful for when a transaction fails execution or when a transaction
+/// has expired in the app's mempool.
+///
+/// The cache places a TTL on the entries, allowing users
+/// to request their transaction to be processed again after the
+/// ttl has expired. This is useful for users who want to re-run a transaction
+/// that failed to execute or to re-add a transaction that has expired.
+#[derive(Clone)]
+pub(crate) struct RemovalCache {
+    cache: HashMap<[u8; 32], (Instant, Arc<anyhow::Error>)>,
+    remove_queue: VecDeque<[u8; 32]>,
+    max_size: NonZeroUsize,
+    time_to_live: Duration,
+}
+
+impl RemovalCache {
+    fn new(max_size: NonZeroUsize, time_to_live: Duration) -> Self {
+        Self {
+            cache: HashMap::new(),
+            remove_queue: VecDeque::with_capacity(max_size.into()),
+            max_size,
+            time_to_live,
+        }
+    }
+
+    /// returns Some(Error) if transaction is cached, where the Error is
+    /// the reason why it was cached. None is returned if the transaction
+    /// is not cached
```

Comment:
```suggestion
    /// the reason why it was cached. None is returned if the transaction
    /// is not cached or is not expired
```

Reply:
I was considering a tx being 'cached' as being both not expired and in the cache, but I can change

Reply:
i see, yeah i got confused from the variable naming, a comment on that would help!

Reply:
sorry i got confused about the tx removal ttl vs the tx in mempool ttl. this makes sense!